website/blog/container-image-streaming.md

```diff
@@ -0,0 +1,70 @@
+
+To support all of these modifications, we added a bit of logic to the snapshotter that attempts to retrieve artifacts from the local filesystem before the remote registry and supports reading the image-to-index mapping from the local filesystem under a special directory, again keyed by the image identifier. Finally, we set up the app and executor to look for all of these artifacts in the cache, generate them if they’re not present, and store them on the local filesystem so the snapshotter can do its thing.
+
+The modified soci-snapshotter is available at [github.com/buildbuddy-io/soci-snapshotter](https://github.com/buildbuddy-io/soci-snapshotter).
+
+![](/img/blog/soci-snapshotter-buildbuddy.webp)
+_The BuildBuddy + soci-snapshotter setup. SOCI artifacts are generated on-demand in the App, stored in the cache, and retrieved by the Executor as needed._
+
+## Authorization
+
+Many of our customer’s container images are password protected in the container registry. We share user-provided registry credentials between the executors and the soci-snapshotter so the snapshotter can fetch registry artifacts as needed, but these credentials expire after a short time period. The setup described here doesn’t permit any unauthorized access to container images or their derived artifacts. The SOCI artifacts are only generated and retrieved for authorized image pull operations, just like standard `podman pull` requests.
+
+## Cache Eviction
+
+We continuously evict the oldest unused artifacts from the BuildBuddy caches to make room for fresh data, and SOCI artifacts are no exception to this. Fortunately, the average last-access-time (including reads) of evicted items for most customers is old enough that only artifacts for rarely pulled container images are affected. Container images used regularly in remote execution or CI only require re-indexing except after being updated in the container registry.
+
+## Bugs!
+
+After initially rolling out lazy image distribution to our executors, we saw some build and test failures due to `Input/output errors` running binaries in containers occasionally. We tracked this down to [a bug](https://github.com/containerd/containerd/issues/6377) in the containerd docker authorization library that was causing it to re-use expired authentication tokens a bit too eagerly. Fortunately, there was a pretty straightforward [workaround](https://github.com/buildbuddy-io/soci-snapshotter/pull/2) while we wait for the [upstream fix](https://github.com/containerd/containerd/pull/8735/files).
```

Comment:
idk if you wanted to also add the race condition you found by turning on the race detector - I thought that was really cool!

Reply:
How could I forget - thanks for the reminder!